flexmeasures/api/common/utils/api_utils.py

```diff
@@ -356,3 +366,64 @@ def get_generic_asset(asset_descriptor, entity_type):
             ea["longitude"],
         )
     return None
+
+
+def save_to_db(
+    timed_values: List[Union[Power, Price, Weather]], forecasting_jobs: List[Job]
+) -> ResponseTuple:
+    """Put the timed values into the database and create forecasting jobs.
+
+    Data can only be replaced on servers in play mode.
+
+    :param timed_values: list of Power, Price or Weather values to be saved
+    :param forecasting_jobs: list of forecasting Jobs for redis queues.
+    :returns: ResponseTuple
+    """
+    current_app.logger.info("SAVING TO DB AND QUEUEING...")
+    try:
+        save_to_session(timed_values)
+        db.session.flush()
+        [current_app.queues["forecasting"].enqueue_job(job) for job in forecasting_jobs]
+        db.session.commit()
+        return request_processed()
+    except IntegrityError as e:
+        current_app.logger.warning(e)
+        db.session.rollback()
+
+        # Allow data to be replaced only in play mode
+        if current_app.config.get("FLEXMEASURES_MODE", "") == "play":
+            save_to_session(timed_values, overwrite=True)
+            [
+                current_app.queues["forecasting"].enqueue_job(job)
+                for job in forecasting_jobs
+            ]
+            db.session.commit()
+            return request_processed()
+        else:
+            return already_received_and_successfully_processed()
+
+
+def determine_belief_horizons(event_values, start, resolution, horizon, prior, sensor):
+    """Determine belief horizons given a horizon, prior, or both,
+    and taking into account the sensor's knowledge horizon function."""
+    event_starts = [start + j * resolution for j in range(len(event_values))]
+    if horizon is not None and prior is not None:
+        belief_horizons_from_horizon = [horizon] * len(event_values)
```

Comment:
Not sure if you like it, but there is (by my current thinking) a way to not repeat both horizon calculations in the code.

We could compute `belief_horizons_from_horizon` and `belief_horizons_from_prior` first, each in their `if` statement (e.g. `if horizon is not None` and return right away if the other one (e.g. `prior` is None).

Then we only check the case if they are both not None (return the max horizons).

If we haven't returned by then, we raise the Exception.

Reply:
You have a real knack for neatly refactoring if-else blocks.